sql/core/src/test/scala/org/apache/spark/sql/DatasetUnpivotSuite.scala

```diff
@@ -0,0 +1,543 @@
+    checkAnswer(unpivoted, longDataWithoutIdRows)
+  }
+
+  test("unpivot with star values") {
+    val unpivoted = wideDataDs.select($"str1", $"str2")
+      .unpivot(
+        Array.empty,
+        Array($"*"),
+        variableColumnName = "var",
+        valueColumnName = "val")
+    assert(unpivoted.schema === StructType(Seq(
+      StructField("var", StringType, nullable = false),
+      StructField("val", StringType, nullable = true))))
+    checkAnswer(unpivoted, longDataWithoutIdRows)
+  }
+
+  test("unpivot with id and star values") {
+    val unpivoted = wideDataDs.select($"id", $"int1", $"long1")
+      .unpivot(
+        Array($"id"),
+        Array($"*"),
+        variableColumnName = "var",
+        valueColumnName = "val")
+
+    assert(unpivoted.schema === StructType(Seq(
+      StructField("id", IntegerType, nullable = false),
+      StructField("var", StringType, nullable = false),
+      StructField("val", LongType, nullable = true))))
+
+    checkAnswer(unpivoted, wideDataDs.collect().flatMap { row => Seq(
+      Row(row.id, "id", row.id),
+      Row(row.id, "int1", row.int1.orNull),
+      Row(row.id, "long1", row.long1.orNull)
+    )})
+  }
+
+  test("unpivot with expressions") {
+    // ids and values are all expressions (computed)
+    val unpivoted = wideDataDs
+      .unpivot(
+        Array(($"id" * 10).as("primary"), $"str1".as("secondary")),
+        Array(($"int1" + $"long1").as("sum"), length($"str2").as("len")),
+        variableColumnName = "var",
+        valueColumnName = "val")
+
+    assert(unpivoted.schema === StructType(Seq(
+      StructField("primary", IntegerType, nullable = false),
+      StructField("secondary", StringType, nullable = true),
+      StructField("var", StringType, nullable = false),
+      StructField("val", LongType, nullable = true))))
+
+    checkAnswer(unpivoted, wideDataDs.collect().flatMap { row =>
+      Seq(
+        Row(
+          row.id * 10,
+          row.str1,
+          "sum",
+          // sum of int1 and long1 when both are set, or null otherwise
+          row.int1.flatMap(i => row.long1.map(l => i + l)).orNull),
+        Row(
+          row.id * 10,
+          row.str1,
+          "len",
+          // length of str2 if set, or null otherwise
+          Option(row.str2).map(_.length).orNull)
+      )
+    })
+  }
+
+  test("unpivot with variable / value columns") {
+    // with value column `variable` and `value`
+    val unpivoted = wideDataDs
+      .withColumnRenamed("str1", "var")
+      .withColumnRenamed("str2", "val")
+      .unpivot(
+        Array($"id"),
+        Array($"var", $"val"),
+        variableColumnName = "var",
+        valueColumnName = "val")
+    checkAnswer(unpivoted, longDataRows.map(row => Row(
+      row.getInt(0),
+      row.getString(1) match {
+        case "str1" => "var"
+        case "str2" => "val"
+      },
+      row.getString(2))))
+  }
+
+  test("unpivot with incompatible value types") {
```

Comment:
This test fails with Scala 2.13:

```
2022-07-30T20:25:08.3651463Z [0m[[0m[0minfo[0m] [0m[0m[31m- unpivot with incompatible value types *** FAILED *** (9 milliseconds)[0m[0m
2022-07-30T20:25:08.3679939Z [0m[[0m[0minfo[0m] [0m[0m[31m  java.lang.IllegalArgumentException: (types, "STRING" (`str1#756021`), "BIGINT" (`long1#756024L`, `long2#756222L`), "INT" (`int1#756023`, `int2#756219`, `int3#756220`, ...)) does not match: "STRING" \(`str1#\d+`\), "INT" \(`int1#\d+`, `int2#\d+`, `int3#\d+`, ...\), "BIGINT" \(`long1#\d+L`, `long2#\d+L`\)[0m[0m
2022-07-30T20:25:08.3688615Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.apache.spark.SparkFunSuite.$anonfun$checkError$2(SparkFunSuite.scala:314)[0m[0m
2022-07-30T20:25:08.3689594Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.apache.spark.SparkFunSuite.$anonfun$checkError$2$adapted(SparkFunSuite.scala:309)[0m[0m
2022-07-30T20:25:08.3690254Z [0m[[0m[0minfo[0m] [0m[0m[31m  at scala.collection.immutable.Map$Map1.foreach(Map.scala:260)[0m[0m
2022-07-30T20:25:08.3690894Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.apache.spark.SparkFunSuite.checkError(SparkFunSuite.scala:309)[0m[0m
2022-07-30T20:25:08.3691621Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.apache.spark.sql.DatasetUnpivotSuite.$anonfun$new$31(DatasetUnpivotSuite.scala:316)[0m[0m
2022-07-30T20:25:08.3692301Z [0m[[0m[0minfo[0m] [0m[0m[31m  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)[0m[0m
2022-07-30T20:25:08.3697872Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)[0m[0m
2022-07-30T20:25:08.3698511Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)[0m[0m
2022-07-30T20:25:08.3699227Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)[0m[0m
2022-07-30T20:25:08.3699809Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.Transformer.apply(Transformer.scala:22)[0m[0m
2022-07-30T20:25:08.3700393Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.Transformer.apply(Transformer.scala:20)[0m[0m
2022-07-30T20:25:08.3701206Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:190)[0m[0m
2022-07-30T20:25:08.3701952Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:204)[0m[0m
2022-07-30T20:25:08.3711793Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:188)[0m[0m
2022-07-30T20:25:08.3712584Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:200)[0m[0m
2022-07-30T20:25:08.3713239Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)[0m[0m
2022-07-30T20:25:08.3714016Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:200)[0m[0m
2022-07-30T20:25:08.3714722Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:182)[0m[0m
2022-07-30T20:25:08.3721520Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:65)[0m[0m
2022-07-30T20:25:08.3722356Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)[0m[0m
2022-07-30T20:25:08.3723258Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)[0m[0m
2022-07-30T20:25:08.3723937Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:65)[0m[0m
2022-07-30T20:25:08.3724639Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:233)[0m[0m
2022-07-30T20:25:08.3725302Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)[0m[0m
2022-07-30T20:25:08.3760716Z [0m[[0m[0minfo[0m] [0m[0m[31m  at scala.collection.immutable.List.foreach(List.scala:333)[0m[0m
2022-07-30T20:25:08.3761569Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)[0m[0m
2022-07-30T20:25:08.3762221Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)[0m[0m
2022-07-30T20:25:08.3762860Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)[0m[0m
2022-07-30T20:25:08.3763534Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:233)[0m[0m
2022-07-30T20:25:08.3780043Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:232)[0m[0m
2022-07-30T20:25:08.3780996Z [0m[[0m[0minfo[0m] [0m[0m[31m  at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1563)[0m[0m
```

https://pipelines.actions.githubusercontent.com/serviceHosts/03398d36-4378-4d47-a936-fba0a5e8ccb9/_apis/pipelines/1/runs/172018/signedlogcontent/23?urlExpires=2022-07-31T10%3A19%3A48.9338315Z&urlSigningMethod=HMACV1&urlSignature=Znam4MgLOGX3zTfeqvN9HR7bwLJ0JiHLbzqQNYjqq%2FA%3D

Reply:
https://github.com/apache/spark/actions/runs/2766863521

Reply:
I'll create a fix.